### ISSUE TITLE
The old mutate the collection while iterating trick

### ISSUE THREAD

**Issue by som-snytt** [scala/bug#12009]
## reproduction steps

- [x] using Scala?

```scala
scala 2.13.2> val xs = collection.mutable.ArrayBuffer(1,2,3)
val xs: scala.collection.mutable.ArrayBuffer[Int] = ArrayBuffer(1, 2, 3)

scala 2.13.2> val it = xs.iterator
val it: Iterator[Int] = <iterator>

scala 2.13.2> it.hasNext
val res0: Boolean = true

scala 2.13.2> xs.clear()

scala 2.13.2> it.hasNext
val res2: Boolean = true

scala 2.13.2> it.next()
val res3: Int = 0

scala 2.13.2> it.to(List)
val res4: List[Int] = List(null, null)

scala 2.13.2> xs.size
val res5: Int = 0
```

## problem

Documentation should tell me not to do that.

It would be nice if the iterator told me there was a problem. It's just using indexes from the view.

See https://github.com/scala/bug/issues/8154

**Comment by joroKr21** [comment]
TIL some countries limit the number of precautionary statements on GHS labels. The EU where I live mandates a maximum of 6. I propose "mutating hazard".

**Comment by SethTisue** [comment]
as I said on Gitter, when someone mutates a collection on which there is a live iterator, I doubt there we can offer any better guarantee _in general_ other than shrug and say "sorry, undefined behavior"

that said, certainly

1) this should be documented somewhere, if it isn't already
2) there's nothing wrong with improving behaviors in particular scenarios (e.g. to throw an exception rather than silently returning `null`) if it can be done without too great an impact to performance or maintainability

**Comment by LPTK** [comment]
@SethTisue do you think it would be acceptable, performance-wise, to have a private `var version: Int` field in mutable collections, which is incremented every time a mutable operation is done, and which is captured by iterators (so iterators can check whether the collection has changed under their feet when doing `hasNext`/`next`).

**Comment by ritschwumm** [comment]
@LPTK this field would probably have to be volatile to ensure the mechanism works accross threads... all in all, this (to me) sounds like another reason to eschwe mutable collections altogether, except in very few and tightly controlled places.

**Comment by Jasper-M** [comment]
The `version` field thing seems to be how the `java.util` collections implements fail-fast iterators.

> this field would probably have to be volatile to ensure the mechanism works accross threads

I don't think mutable collections offer any synchronization guarantees in general anyway.

**Comment by ritschwumm** [comment]
well, they don't offer any guarantees concerning mutation while using a live iterator either, and still people seem to be surprised when that goes wrong :)

**Comment by LPTK** [comment]
@ritschwumm IMHO there's a huge difference between behaving in a buggy way in single-threaded contexts, and behaving in a buggy way in multi-threaded contexts. It's widely accepted that the latter is expected of a non-thread-safe API; but the former just seems to reflect a bad API/implementation design.

I think Java got at least that part right: basic Java collections are not thread-safe, but they do make sure you're not misusing them in single-threaded code (if they do it, I guess it shows it's not prohibitive in terms of performance).

**Comment by SethTisue** [comment]
perhaps my initial reaction was too pessimistic/dismissive?

> do you think it would be acceptable, performance-wise

@LPTK if the Java collections do that and then that's good evidence it would be okay for us performance wise

do the checks they do seem simple/maintainable? or does it get hairy? if not, that would be also be evidence this could be feasible for us

**Comment by ritschwumm** [comment]
@LPTK yeah, agreed - with multi-threading, people tend to know it won't work well with mutable collections in the general case.

i wonder why, though - to me, an iterator is not so much different than a suspended thread. if you ignore memory model issues like cpu caches, the code ran between one call to an iterator and the next is conceptually the same as running something in a separate thread. if that mucks with the state the iterator relies on, you are in trouble...

the more interesting question to me is - is there a good way to integrate a check for this in the current colllections implementation? ideally, it would be easier for someone working on the implementation of a mutable collection to have the version field updated for each and every mutation than not doing it. the same goes for actually checking that field in an iterator created on that collection.

personally, i'd prefer collections not doing any of this over collections where some mutations are checked, and others are not, and there's no way to know which ones are. i already see future me having to refactore some code where a junior _relied_ on version mismatch exceptions to be thrown in very specific cases...

**Comment by NthPortal** [comment]
I believe (some of?) the Java collections detect concurrent modification across multiple threads as well - see https://docs.oracle.com/javase/8/docs/api/java/util/ConcurrentModificationException.html

**Comment by retronym** [comment]
Some more verbiage from Java's documentation for `ArrayList`

```
 The iterators returned by this class's iterator and listIterator methods are fail-fast: if the list is structurally modified at any time after the iterator is created, in any way except through the iterator's own remove or add methods, the iterator will throw a ConcurrentModificationException. Thus, in the face of concurrent modification, the iterator fails quickly and cleanly, rather than risking arbitrary, non-deterministic behavior at an undetermined time in the future.

Note that the fail-fast behavior of an iterator cannot be guaranteed as it is, generally speaking, impossible to make any hard guarantees in the presence of unsynchronized concurrent modification. Fail-fast iterators throw ConcurrentModificationException on a best-effort basis. Therefore, it would be wrong to write a program that depended on this exception for its correctness: the fail-fast behavior of iterators should be used only to detect bugs. 
```

**Comment by NthPortal** [comment]
the only simple way to apply this "in bulk" to most of the collections without lots of code duplication and manual work is to add a `MutationTracking` trait with a `mutationCount` field (or similar), a method to increment that count that can be dropped in at the beginning/end of every mutating method, and an inner class/trait `CheckedIterator` that checks for a change to the mutation count in its methods, and throws an exception if that is found. sadly, we cannot add `MutationTracking` to public collection implementations, as that would require it to be public as well, thus breaking bincompat.

Perhaps we could have a `MutationTracker` that is a field/member of collections though (and is `private[collection]`, thus not breaking bincompat)?

**Comment by Ichoran** [comment]
Bincompat aside, I am not convinced that fail-fast-on-concurrent-modification is a wise strategy.  If you have two references to the mutable collection itself, you're not guaranteed to get sensible results.  Indeed, you fundamentally can't write anything that will do the job (absent some Rust-like ownership system).  That you _can_ do the job (in part) with `Iterator` doesn't mean that you _should_.  You already have to do the job manually with the collection itself; so do the job manually with the iterator, too.

Furthermore, iterators themselves are a pile of unintended concurrent modifications waiting to happen.  You have to handle those manually too.

So requiring increased memory usage and computation for almost every operation where iterators are used, while leaving undetected (and undetectable) problems of a very similar nature--and which, if solved manually, will also solve this problem--seems a bad idea to me.  The exception would be if we knew that (1) this kind of error is common overall when using iterators, and also (2) the other errors of using mutable collections concurrently or the same iterator concurrently are rare in comparison.

Instead, I think we should spot-check for any particularly perverse behavior and fix it.  Getting nulls in the `ArrayBuffer` iterator is arguably that; the solution in this case (hopefully not too performance-damaging) is for iterators on mutable collections to use the length of the collection rather than caching the length.  It's a sensible thing to do for immutable collections, but not as sensible for mutable ones.

**Comment by NthPortal** [comment]
> If you have two references to the mutable collection itself, you're not guaranteed to get sensible results.

that is not my understanding of the collections' semantics. my understanding is that if you have two references to a mutable collection, and use each reference interchangeably at the same time, you'll be fine.

Perhaps we're referring to different things, in which case some of my further comments may be irrelevant. I'm referring to a concurrency-unaware detection, that is not guaranteed to fail with concurrent modification across different threads, but should reliably fail if a single thread modifies a collection while using its iterator.

> iterators themselves are a pile of unintended concurrent modifications waiting to happen.

this, unfortunately, is a bit more difficult to solve, but is definitely an issue that could use both better documentation and research into possible solutions

> So requiring increased memory usage

I don't think it's significant - it's a fixed and small cost per collection

> So requiring increased [...] computation

assuming we're only doing a best-effort detection, and not perfect concurrent detection, I think it will be a lot cheaper than you think. of course, the only way to know is benchmarking

>  The exception would be if we knew that (1) this kind of error is common overall when using iterators

I have potentially discovered a compiler bug (investigating now) after changing only `ListBuffer` to have a fail-fast iterator, so I expect it's a lot more common than we think

> I think we should spot-check for any particularly perverse behavior and fix it.

I think it is actually significantly easier to just implement checks for all collection, which will be generally safer across the board as well

> It's a sensible thing to do for immutable collections, but not as sensible for mutable ones.

unless I'm misunderstanding something, immutable collections can't be modified, so they don't need any checks

**Comment by retronym** [comment]
> If you have two references to the mutable collection itself, you're not guaranteed to get sensible results.

The typical case that the fail-fast behaviour flags is mutation inside a loop.

```
for (x <- xs) {
  if (condition) xs += 1
}
```

I think its worth the effort to implement this. IMO the overhead is acceptably low in most cases; one notable exception would be implementing this for a mutable doubly linked list, but in that case one can implement a forward iterator to handle mutation correctly anyway.

**Comment by NthPortal** [comment]
> I think its worth the effort to implement this.

I would like to add that, due to adding a fail-fast iterator *just* to `ListBuffer`, I have discovered several bugs, including in appending a `ListBuffer` to itself. I think that shows the value in this endeavour.

**Comment by Ichoran** [comment]
Extra attention on code is always useful for finding bugs, so that's good!

But I don't understand what the conceptual difference is between these two:

```
for (x <- xs.iterator) { if (p(x)) xs += f(x) }
for (x <- xs)          { if (p(x)) xs += f(x) }
```

It's the same problem.  Are we going to fix both?  If not, is the former considerably more common?

Anyway, I think benchmarking will be important.  I would expect fast operations like

```scala
var n = 0
for (x <- xs.iterator) n += x.length
```

to slow down by about 20-30% except in cases where it's fully inlined and the extra checks are removed.  Iterators also usually aren't very big to begin with, so if their copy of the mutation count doesn't fit in unused space in the allocation, the size change could be something like 32-byte iterators becoming 40 bytes (25% increase).

I would also expect ordinary loop operations like

```scala
val ys = collection.mutable.ArrayBuffer.empty[String]
for (x <- xs) ys += x
```

to be slowed down by about 5-10% as the `mutationCount` gets continually increased.

If mutating in place is considered a mutation (which is necessary for any kind of buffering in iterators to work properly), then I would expect to see also ~20-30% slowdowns on things like `xs(i) = "salmon"`.

These rough hunches should be verified, of course.

**Comment by NthPortal** [comment]
> But I don't understand what the conceptual difference is between these two:
>
> ```scala
> for (x <- xs.iterator) { if (p(x)) xs += f(x) }
> for (x <- xs)          { if (p(x)) xs += f(x) }
> ```
>
> It's the same problem. Are we going to fix both? If not, is the former considerably more common?

honestly, I expect both of those examples to go equally poorly. what I'm fixing is `xs ++= xs`.

**Comment by Ichoran** [comment]
@NthPortal - If there is a fix for `xs ++= xs` breaking, that's good.  But that's a different issue than general-purpose fail-fast behavior.  So I think the point is still relevant--we should either care about both or neither, unless one form is a common mistake and the other is rare.

**Comment by NthPortal** [comment]
I think you could make a justifiable claim that `xs.foreach(xs += _)` is invalid/bad code, but I don't think you can for `xs ++= xs`

**Comment by Ichoran** [comment]
@NthPortal - I can't?  Now I have to try!  :rofl:

`xs ++= ys` is expected to be equivalent to `ys.foreach(xs += _)`.  Therefore, `xs ++= xs` and `xs.foreach(xs += _)` should have the same behavior.

Is that justifiable enough?

**Comment by LPTK** [comment]
> But I don't understand what the conceptual difference is between these two:
> 
> ```scala
> for (x <- xs.iterator) { if (p(x)) xs += f(x) }
> for (x <- xs)          { if (p(x)) xs += f(x) }
> ```
> 
> It's the same problem. Are we going to fix both?

My understanding was that both would be fixed.

When I first raised the problem in the Scala gitter, which lead to this issue being created, I was looking specifically at iterators. But doing the checks within `foreach` is just as easy to implement (using the same `version` field trick), so it should also be done.

**Comment by NthPortal** [comment]
> My understanding was that both would be fixed.

depends what you mean by fixed. if you mean that both will throw an exception, then yes, that's doable (and since `foreach` is by default implemented with `iterator`, it comes free). if you mean that they will work, I don't think that's feasible

**Comment by LPTK** [comment]
Yes, by "fixed" I meant that they'd throw the `ConcurrentModification` exception.

NB: There could also be collections whose contract says single-threaded concurrent modifications are allowed for certain operations. It's probably okay to stop the iteration early if some elements of an `ArrayBuffer` are removed/cleared in the middle of iteration. But for sure it shouldn't start returning `null`s or other nonsensical results — it should be either fail-fast behavior or sensible behavior, and the behavior should be properly documented either way.

**Comment by Ichoran** [comment]
I'm not sure it's a good idea to change the runtime behavior of either of these, though.  If it already didn't work (it didn't hang, but did something weird), then sure, an exception is better.

But if it _does_ work already, then throwing an exception _instead_ is a bad idea.

If the length was cached, then making it throw an exception prevents `xs.foreach(xs += _)` from doubling the contents of the list.

If the length is not cached, then making it throw an exception prevents a one-liner recursive-like file traversal via
`xs.foreach(x => if (x.isDirectory) xs ++= x.listFiles)`

I don't think we're allowed to clobber working code like that _especially_ in a point release, _especially_ as a runtime error.  We don't promise either behavior, but we do promise not to randomly break working code.

**Comment by NthPortal** [comment]
I don't think we should be tied to UB, point release or not.

> prevents a one-liner recursive-like file traversal via `xs.foreach(x => if (x.isDirectory) xs ++= x.listFiles)`

you can do it with 2 lines without mutating under yourself:

```scala
val queue = mutable.Queue.from(xs)
// why is there no `dequeueOption()`?
Iterator.continually(if (queue.isEmpty) None else queue.dequeue()).takeWhile(_.isDefined).tapEach(x => if (x.isDirectory) queue.enqueueAll(x.listFiles))
```

despite the fact that it's poorly documented if at all, mutating a collection while iterating over it has *never* been supported. I don't think we should be bound to it just because it happened to work so far without blowing up in some cases.

**Comment by LPTK** [comment]
@Ichoran I'm actually okay with making some collections keep their old behavior if that was sane behavior (which should then be documented).

It looks like currently for `ArrayBuffer` the size is cached so `xs.foreach(xs += _)` for doubling the collection works, but the file-traversal won't work.

At any rate, something needs to be done for the case when elements are removed from the buffer making its size less than the cached size, which _never_ worked properly and could be causing bugs as we speak.

In that case, either throw an exception or stop the iteration early. Not sure which to prefer, but the latter will avoid some runtime exceptions (at the price of not detecting some actual errors).

**Comment by SethTisue** [comment]
can y'all who are interested in this review https://github.com/scala/scala/pull/9174 ...?
